File: lesson-4/java/MrCoxallStack.java
/*
 * The MrCoxallStack class.
 *
 * @author Mohamad T
 * @version 1.0.0
 * @since 2024-10-16
 */

import java.util.ArrayList;
import java.util.List;

/**
 * The MrCoxalStack class.
 */
public class MrCoxallStack {
    /**
     * The list used for the stack.
     */
    private final List<String> stackList = new ArrayList<>();

    /**
     * The showStack method to show the stored values of the stack in a string.
     *
     * @return a string containing the values of the stack.
     */
    public String showStack() {
        String stackValues = "";
        for (int counter = 0; counter < this.stackList.size(); counter++) {
            stackValues += this.stackList.get(counter) + ", ";
        }
        stackValues = stackValues.substring(0, stackValues.length() - 2);

        return stackValues;
    }

    /**
     * The addStr method to append a string to the stack.
     *
     * @param str A string that you would like to append.
     */
    public void pushStr(String str) {
        this.stackList.add(str);
    }

    /**
     * The popItem method to remove the top item and return it in a string.
     *
     * @return The top item in a string
     */
    public String popItem() {
        // Get the size of the list
        final int stackLength = this.stackList.size();
        String outputString;
        if (stackLength < 1) {
            // Return this if it is empty
            outputString = "The stack is empty!";
        }
        // Pick the top item
        outputString = this.stackList.get(stackLength - 1);

        // Remove item from list
        this.stackList.remove(stackLength - 1);

        return outputString;
    }

    /**
     * The isEmpty method to check if the stack contains anything.
     *
     * @return A true or false in a string
     */
    public String isEmpty() {
        final boolean empty;
        if (this.stackList.size() < 1) {
            empty = true;
        } else {
            empty = false;
        }

        final String boolAsString = Boolean.toString(empty);

        return boolAsString;
    }

    /**
     * The size method to check the size of the stack.
     *
     * @return An integer value of the stack's size
     */
    public int size() {
        return this.stackList.size();
    }
}

File: lesson-4/ts/MrCoxallStack.ts
/*
 * This is the MrCoxallStack class.
 *
 * Author: Mohamad T
 * Version: 1.0
 * Since: 2024-10-16
 */

export class MrCoxallStack {
  /**
   * The MrCoxallStack list
   */
  private readonly stackList: string[] = []

  /**
   * This method returns the stack as a string.
   *
   * @return {string} The stack as a string.
   */
  public get showStack (): string {
    let stackValues = ''
    for (let counter = 0; counter < this.stackList.length; counter++) {
      stackValues = stackValues + this.stackList[counter] + ', '
    }
    stackValues = stackValues.substring(0, stackValues.length - 2)

    return stackValues
  }

  /**
   * This method removes the top item and returns it as a string.
   *
   * @return {string}
   */
  public get popItem (): string {
    // Ensure that the stack is full.
    if (this.stackList.length < 1) {
      return 'The stack is empty!'
    }

    // Grab top item
    const topItemIndex = this.stackList.length - 1
    const topItem = this.stackList[topItemIndex]

    // Remove the top item
    this.stackList.splice(topItemIndex, 1)

    return topItem
  }

  /**
   * This method returns a string that is either "true" or "false"
   *  to check if the stack is empty or not.
   *
   *  @return {string} True or False
   */
  public get isEmpty (): string {
    let stackEmpty: boolean
    if (this.stackList.length < 0) {
      stackEmpty = true
    } else {
      stackEmpty = false
    }
    return stackEmpty.toString()
  }

  /**
   * This method gets the size of the stack
   *
   * @return {number}
   */
  public get size (): number {
    return this.stackList.length
  }

  /**
   * This method pushes a string into the stack.
   *
   * @param {string} pushedStr - The string to push into the stack.
   * @return {void}
   */
  public pushStr (pushedStr: string): void {
    this.stackList.push(pushedStr)
  }

  /**
   * This method shows the top item without changing anything.
   *
   * @return {string}
   */
  public get peekItem (): string {
    if (this.stackList.length < 1) {
      return 'The stack is empty!'
    }

    // Grab top item
    const topItemIndex = this.stackList.length - 1

    // Return top item
    return this.stackList[topItemIndex]
  }

  /**
   * This method clears the stack and returns nothing.
   *
   * @return {void}
   */
  public clearStack (): void {
    while (this.stackList.length > 0) {
      this.stackList.splice(-1, 1)
    }
  }
}
